src/Lecture2.hs

```diff
@@ -207,8 +249,19 @@ The algorithm of merge sort is the following:
 >>> mergeSort [3, 1, 2]
 [1,2,3]
 -}
+
+half :: [a] -> ([a], [a])
+half xs =
+    ((take s xs), drop s xs)
+    where
+        s = (length xs ) `div` 2
```

Comment:
You can use `splitAt` instead of `take + drop` to avoid traversing the first half twice.

Additionally, you can avoid using `length` if you notice that you don't necessarily need to split in the middle. You only need to split it into two lists of approximately the same size.

In that case, you can implement a solution that traverses a list only once and splits its elements into two lists of elements on only even positions and only odd positions.

Alternatively, you can implement a bottom-up approach by converting a list of elements to a list of of singleton lists and merging them pairwise until only one list is left.